### PR TITLE
Operator `[]` for `DynamicType`

### PR DESCRIPTION
**Pull request by zasdfgbnm**
This PR is stacked on https://github.com/NVIDIA/Fuser/pull/523, it adds support for `operator []` for `DynamicType`.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build

File: test/test_dynamic_type.cpp [hunk 1]
```diff
@@ -602,20 +602,16 @@ TEST_F(DynamicTypeTest, ExamplesInNote) {
...
-        1);
-    EXPECT_EQ(
-        y.as<std::list<IntFloatVecList>>()
-            .back()
-            .as<std::vector<IntFloatVecList>>()
-            .back()
-            .as<float>(),
+            .as<std::vector<IntFloatVecList>>()[1],
         2.0f);
+    EXPECT_THAT(
```
Comment (zasdfgbnm) [inline test/test_dynamic_type.cpp:611]: `std::list` can not be indexed

**Comment by zasdfgbnm** [comment]
!build

**Review by naoyam — APPROVED** [review]
LGTM

**Pull request merged**